src/main/java/org/opensearch/security/configuration/SecurityFlsDlsIndexSearcherWrapper.java

```diff
@@ -103,36 +113,125 @@ public SecurityFlsDlsIndexSearcherWrapper(
+        );
     }
 
     @SuppressWarnings("unchecked")
     @Override
     protected DirectoryReader dlsFlsWrap(final DirectoryReader reader, boolean isAdmin) throws IOException {
-
         final ShardId shardId = ShardUtils.extractShardId(reader);
         PrivilegesEvaluationContext privilegesEvaluationContext = this.dlsFlsBaseContext.getPrivilegesEvaluationContext();
+        final String indexName = (shardId != null) ? shardId.getIndexName() : null;
 
         if (log.isTraceEnabled()) {
-            log.trace("dlsFlsWrap(); index: {}; privilegeEvaluationContext: {}", index.getName(), privilegesEvaluationContext);
+            log.trace("dlsFlsWrap(); index: {}; isAdmin: {}", indexName, isAdmin);
         }
 
+        // 1. If user is admin, or we have no shard/index info, just wrap with default logic (no doc-level restriction).
         if (isAdmin || privilegesEvaluationContext == null) {
-            return new DlsFlsFilterLeafReader.DlsFlsDirectoryReader(
-                reader,
-                FieldPrivileges.FlsRule.ALLOW_ALL,
-                null,
-                indexService,
-                threadContext,
-                clusterService,
-                auditlog,
-                FieldMasking.FieldMaskingRule.ALLOW_ALL,
-                shardId,
-                metaFields
-            );
+            return wrapWithDefaultDlsFls(reader, shardId);
         }
 
+        assert !Strings.isNullOrEmpty(indexName);
+        // 2. If resource sharing is disabled or this is not a resource index, fallback to standard DLS/FLS logic.
+        if (!this.isResourceSharingEnabled || !OpenSearchSecurityPlugin.getResourceIndices().contains(indexName)) {
+            return wrapStandardDlsFls(privilegesEvaluationContext, reader, shardId, indexName, isAdmin);
+        }
+
+        // TODO see if steps 3,4,5 can be changed to be completely asynchronous
+        // 3.Since we need DirectoryReader *now*, we'll block the thread using a CountDownLatch until the async call completes.
+        final AtomicReference<Set<String>> resourceIdsRef = new AtomicReference<>(Collections.emptySet());
+        final AtomicReference<Exception> exceptionRef = new AtomicReference<>(null);
+        final CountDownLatch latch = new CountDownLatch(1);
+
+        // 4. Perform the async call to fetch resource IDs
+        this.resourceAccessHandler.getAccessibleResourceIdsForCurrentUser(indexName, ActionListener.wrap(resourceIds -> {
+            log.debug("Fetched resource IDs for index '{}': {}", indexName, resourceIds);
+            resourceIdsRef.set(resourceIds);
+            latch.countDown();
+        }, ex -> {
+            log.error("Failed to fetch resource IDs for index '{}': {}", indexName, ex.getMessage(), ex);
+            exceptionRef.set(ex);
+            latch.countDown();
+        }));
+
+        // 5. Block until the async call completes
         try {
+            latch.await();
```

Comment:
Blocking such threads is dangerous. See comment at https://github.com/opensearch-project/security/pull/5016/files#r1924173306

Reply:
https://github.com/opensearch-project/security/pull/5016#discussion_r1927298282.